pxtcompiler/emitter/decompiler.ts

```diff
@@ -2409,20 +2400,20 @@ ${output}</xml>`;
             return undefined;
         }
 
-        function checkVariableDeclaration(n: ts.VariableDeclaration, env: DecompilerEnv) {
-            let check: string;
-
+        function checkVariableDeclaration(n: ts.VariableDeclaration, env: DecompilerEnv): string | undefined {
             if (n.name.kind !== SK.Identifier) {
-                check = Util.lf("Variable declarations may not use binding patterns");
+                return Util.lf("Variable declarations may not use binding patterns");
             }
-            else if (!n.initializer) {
-                check = Util.lf("Variable declarations must have an initializer");
+            if (!n.initializer) {
+                return Util.lf("Variable declarations must have an initializer");
             }
-            else if (!isAutoDeclaration(n)) {
-                check = checkExpression(n.initializer, env);
+            if (isAutoDeclaration(n, env)) {
+                return undefined;
             }
-
-            return check;
+            if (n.type && !isBlocklyType(n.type)) {
```

Comment:
This check does not seem right to me; MakeCode Arcade has other types that we decompile (sprites, images) and you can get other decompilable types based on the extensions in a project (neopixel has the Strip class, for example).

Reply:
If a variable declaration has an explicit type definition and the type is not `boolean`, `number`, `string`, an array or a jagged array of any of these types, it will be converted to gray block.

Variable declaration without explicit type definition will be converted to normal blocks if possible.

The idea behind is that when the programmer explicitly defines the type when declaring a variable, he/she really means it and such information should not be erased during blocks and JavaScript conversions.

Initially `boolean`, `number`, `string` and their arrays and arrays of arrays are not exempted, but I found that Blockly Compiler generated variable declarations with explicit type definition sometimes.

Reply:
Here is a test case that illustrates my point; this decompiles correctly on the beta of makecode.microbit.org but decompiles to a grey-block in the editor link you provided.

1. Create a new project
1. Switch to JavaScript and add a `custom.ts` file (the + button above the file explorer)
1. In that `custom.ts` file, paste this code:
    ```typescript
    /**
    * Custom blocks
    */
    //% weight=100 color=#0fbc11 icon=""
    namespace custom {
        export class TestClass {
    
        }
    
        //% block
        export function typeTest(): TestClass {
            return new TestClass();
        }
    }
    ```

1. Switch back to main.ts and paste this code:
    ```typescript
    let test: custom.TestClass = custom.typeTest()
    ```
1. Try to decompile

Reply:
Yes, this is the intended behavior of this PR.

`let test: custom.TestClass = custom.typeTest()` is decompiled to gray blocks while `let test = custom.typeTest()` is decompiled to normal blocks.

Reply:
It shouldn't though; our policy is to decompile everything as best we can to minimize the number of grey blocks.

Reply:
Do you think it is acceptable to always generate gray blocks for the following types?

- TupleType
- UnionType
- IntersectionType
- ParenthesizedType

I think when the programmer uses these types, the type definition should be kept during conversions.

Reply:
I think I see your point, but if the type they used is concrete (not an expression/combination) and matches the type that we would emit we should not generate a grey block.

I agree that for more complex types like the ones you mention we should definitely give a grey block.

Reply:
I have changed the check to look for advanced types and added test cases for those types. The GitHub page has also been updated.